docs/initialization-steps-and-the-related-tracer.org

```diff
@@ -0,0 +1,624 @@
+#+END_SRC
+
+    There few events that are relevant to ~ChainDB.openDB~
+    + ~TraceOpenEvent~
+    + ~TraceImmutableDBEvent~
+    + ~TraceVolatileDBEvent~
+    + ~TraceLedgerEvent~
+    + ~TraceLedgerReplayEvent~
+    + ~TraceInitChainSelEvent~
+
+    Below we look at each event more closely.
+
+*** ~TraceOpenEvent~ - a.k.a opening events
+
+#+BEGIN_SRC haskell
+   | TraceOpenEvent              (TraceOpenEvent               blk)
+#+END_SRC
+
+     Constructor ~TraceOpenEvent~ wraps a ~TraceOpenEvent blk~. Those events are
+     fired when particular DB is opened.
+
+    #+BEGIN_SRC haskell
+data TraceOpenEvent blk =
+    -- | The ChainDB was opened.
+    OpenedDB
+      (Point blk)  -- ^ Immutable tip
+      (Point blk)  -- ^ Tip of the current chain
+
+    -- | The ChainDB was closed.
+  | ClosedDB
+      (Point blk)  -- ^ Immutable tip
+      (Point blk)  -- ^ Tip of the current chain
+
+    -- | The ImmutableDB was opened.
+  | OpenedImmutableDB
+      (Point blk)          -- ^ Immutable tip
+      ImmutableDB.ChunkNo  -- ^ Chunk number of the immutable tip
+
+    -- | The VolatileDB was opened.
+  | OpenedVolatileDB
+
+    -- | The LedgerDB was opened.
+  | OpenedLgrDB
+  deriving (Generic, Eq, Show)
+    #+END_SRC
+
+*** ~TraceImmutableDBEvent~ - a.k.a ~ImmutableDB~'s events
+
+#+BEGIN_SRC haskell
+   | TraceImmutableDBEvent       (ImmutableDB.TraceEvent       blk)
+#+END_SRC
+
+     ImmutableDB has its own set of events defined at ~Ouroboros.Consensus.Storage.ImmutableDB.Impl.Types~
+
+     #+BEGIN_SRC haskell
+ data TraceEvent blk =
+     NoValidLastLocation
+   | ValidatedLastLocation ChunkNo (Tip blk)
+     -- Validation of previous DB
+   | ValidatingChunk  ChunkNo
+   | MissingChunkFile ChunkNo
+   | InvalidChunkFile ChunkNo (ChunkFileError blk)
+   | ChunkFileDoesntFit (ChainHash blk) (ChainHash blk)
+     -- ^ The hash of the last block in the previous epoch doesn't match the
+     -- previous hash of the first block in the current epoch
+   | MissingPrimaryIndex   ChunkNo
+   | MissingSecondaryIndex ChunkNo
+   | InvalidPrimaryIndex   ChunkNo
+   | InvalidSecondaryIndex ChunkNo
+   | RewritePrimaryIndex   ChunkNo
+   | RewriteSecondaryIndex ChunkNo
+   | Migrating Text
+     -- ^ Performing a migration of the on-disk files
+
+     -- Delete after
+   | DeletingAfter (WithOrigin (Tip blk))
+     -- Closing the DB
+   | DBAlreadyClosed
+   | DBClosed
+     -- Events traced by the index cache
+   | TraceCacheEvent !TraceCacheEvent
+   deriving (Eq, Generic, Show)
+     #+END_SRC
+
+     Because we have a constructor ~TraceImmutableDBEvent ::
+     ImmutableDB.TraceEvent -> ChainDB.TraceEvent~  we can define a tracer for
+     ImmutableDB's events with a ~contramap~ from ~Data.Functor.Contravariant~
+     using existing ~cdbTracer~ (from ~ChainDbArgs~).
+
+     #+BEGIN_SRC haskell
+         , immTracer           = contramap TraceImmutableDBEvent cdbTracer
+     #+END_SRC
+
+     There are few events worth noticing
```

Comment:
```suggestion
     There are a few events worth noticing
```

There are other occurrences of "there are few", should we instead write "there are a few"?